cassandra/tablets.py

```diff
@@ -0,0 +1,99 @@
+# Experimental, this interface and use may change
+from threading import Lock
+
+class Tablet(object):
+    """
+    Represents a single ScyllaDB tablet.
+    """
+    first_token = 0
+    last_token = 0
+    replicas = None
+
+    def __init__(self, first_token = 0, last_token = 0, replicas = None):
+        self.first_token = first_token
+        self.last_token = last_token
+        self.replicas = replicas
+
+    @staticmethod
+    def _is_valid_tablet(replicas):
+        return replicas is not None and len(replicas) != 0
+
+    @staticmethod
+    def from_row(first_token, last_token, replicas):
+        if Tablet._is_valid_tablet(replicas):
+            tablet = Tablet(first_token, last_token,replicas)
+            return tablet
+        return None
+
+# Experimental, this interface and use may change
+class Tablets(object):
+    _lock = None
+    _tablets = {}
+
+    def __init__(self, tablets):
+        self._tablets = tablets
+        self._lock = Lock()
+    
+    def get_tablet_for_key(self, keyspace, table, t):
+        tablet = self._tablets.get((keyspace, table), [])
+        if tablet == []:
+            return None
+        
+        id = bisect_left(tablet, t.value, key = lambda tablet: tablet.last_token)
+        if id < len(tablet) and t.value > tablet[id].first_token:
+            return tablet[id]
+        return None
+
+    def add_tablet(self, keyspace, table, tablet):
+        with self._lock:
+            tablets_for_table = self._tablets.setdefault((keyspace, table), [])
+
+            start = None
+            insert_idx = bisect_left(tablets_for_table, tablet.first_token, key = lambda t: t.first_token)
+            if insert_idx > 0 and insert_idx - 1 < len(tablets_for_table) and tablets_for_table[insert_idx - 1].last_token > tablet.first_token and tablets_for_table[insert_idx - 1].first_token < tablet.last_token:
```

Comment:
Why is `insert_idx - 1 < len(tablets_for_table)` condition required? If you tried to access `tablets_for_table[insert_idx + 1]` somewhere then I'd make sens, but right now I don't understand it.

Reply:
Now that I think of it, how can it ever be false? Afaik bisect_left can't return value greater than len of list